Unified webfonts

johnspackman opened the pull request:
Fixes https://github.com/qooxdoo/qooxdoo/issues/10494

The tl;dr of this PR is that:

- Manifest.json's `provides.webfonts` is deprecated and now `provides.fonts` is preferred
- The entries in `provides.fonts` are declaring available fonts "IDs", which could be reusable private names (eg `qx.theme.monospace`) or ones which follow the well-known name (eg `MaterialIcons`)
- Classes declare that they need a font by using `@usefont(FontIdFromManifest)`
- The font can be defined with URLs (for use with a CDN) and/or with sources
- The compiler decides which font files (if any) to include in the application build, and the user can decide whether to use the CDN or compile fonts into the app (via command line and `compile.json` configuration)
- Glyphs are manually extracted via `qx export-glyphs`, instead of on every compile

The benefits of all this is faster compiles, with smaller build targets - and no space consumed by icon fonts at all

johnspackman committed:
fixes https://github.com/qooxdoo/qooxdoo/issues/10494

johnspackman committed:
fixes

johnspackman committed:
lint and other fixes

johnspackman committed:
Add support for Roboto via CDN; bugfix

derrell commented on docs/desktop/gui/theming.md:
````diff
@@ -236,74 +236,62 @@ It is important to note that you can only specify values available as property
+      },
 ```
 
+In the first example ("qx.theme.monospace") we're defining a short cut that says that anywhere that the font name "qx.theme.monospace" is used, Qooxdoo
````
shortcut (all one word)

derrell commented on docs/desktop/gui/theming.md:
```diff
@@ -236,74 +236,62 @@ It is important to note that you can only specify values available as property
+should switch to using the font family in the array.  In this case, the various Qooxdoo themes often use this one name, and it makes it a touch easier to
+update it in the future.
+
+In the second example ("Roboto"), we're describing a Google font that is stored inside Qooxdoo repo - it's a copy of the Google font, and will be compiled
```
inside *the* Qooxdoo repo

derrell commented on source/class/qx/theme/manager/Font.js:
```diff
@@ -243,6 +247,15 @@ qx.Class.define("qx.theme.manager.Font", {
+     * @internal subject to change
+     * @return {Object}
+     */
+    getManifestFonts() {
```
If this is internal, should it be public? Maybe protected? Or prefixed with '$'?

derrell commented on source/class/qx/tool/compiler/app/Library.js:
```diff
@@ -366,6 +385,23 @@ qx.Class.define("qx.tool.compiler.app.Library", {
+    /**
+     * Detects whether the filename is one of the library's fonts
+     *
+     * @param {*} filename
```
`filename.endsWith` is used, so `filename` must be of type String.

derrell commented on source/class/qx/tool/compiler/app/ManifestFont.js:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @param target  {qx.tool.compiler.targets.Target} the target
+     * @param application  {qx.tool.compiler.app.Application} the application being built
+     * @return {String}
```
`localFonts` and `sources` are not documented

johnspackman replied:
github anomaly - you have to go back to the files tab and find the file again, when you click on the link above it takes you to the original commit instead of the updated file

derrell replied:
Ah, got it. Thanks.

oetiker commented:
can a package provide pre-extracted glyphes ?

oetiker commented:
Is this backward compatible in the sense that we can put it into 7.6 ?

johnspackman commented:
> can a package provide pre-extracted glyphes ?

yes - the glyph files and the source files are expected to in the resources of the library that declares them

> Is this backward compatible in the sense that we can put it into 7.6 ?

yes, it should be because the `provides.webfonts` is not changed.  You should get a deprecated warning when the compiler see you use the old `provides.webfonts`

johnspackman commented on source/class/qx/tool/compiler/resources/ImageLoader.js:
```diff
@@ -50,18 +50,8 @@ qx.Class.define("qx.tool.compiler.resources.ImageLoader", {
      * @Override
```
hmm - no i dont think it is, but it's something i do habitually to explain that this function does not require documentation / that the super class function documents it

johnspackman commented on source/class/qx/tool/compiler/app/ManifestFont.js:
```diff
@@ -0,0 +1,148 @@
+    },
+
+    /**
+     * Return bootstrap code that is executed before the Application starts.
```
not that I can think of without an external file that uses templates, which isnt very practical.  The compiler is a code generator after all.  This code came from the `qx.tool.compiler.app.WebFont` and there is plenty of other places where code is generated like this

johnspackman committed:
no message

hkollmann committed:
Merge branch 'master' into unified-webfonts

hkollmann commented:
Is it possible to provide a migration class for 7_5_6?

oetiker commented:
how about adding support for https://fontsource.org/ ?

johnspackman committed:
migration

johnspackman committed:
Merge branch 'unified-webfonts' of github.com:johnspackman/qooxdoo into unified-webfonts

* 'unified-webfonts' of github.com:johnspackman/qooxdoo:
  "Release v7.5.1"
  revert
  revert
  Update Manifest.json
  "Release v7.5.1"
  Update release-notes.7.0.md
  Update release-notes.7.0.md
  Update release-notes.7.0.md
  long running compile processes do  not exit (#10539)
  Fix update notifier in qx starter (#10543)
  relaod config. We need a fresh model here because data will be verified. (#10542)
  fix wss address in test (#10541)

johnspackman commented:
> Is it possible to provide a migration class for 7_5_6?

done

> how about adding support for https://fontsource.org/ ?

I've just tried doing this, but i'm not sure how to implement it - eg I tried with their example `npm install @fontsource/open-sans`.

AIUI the web page would need to add at least `./node_modules/@fontsource/open-sans/index.css`, which references 16 font files as CSS `url(...)`s; there are other variants that the user could choose (eg different weights, decoration, or alphabet).  The font files exist in `./node_modules/@fontsource/open-sans/files/`, where there are 237 files.

Unless we expect that the web server will serve the entire of `node_modules/@fontsource`, we would have to copy `node_modules/@fontsource` into the build target directory, and then we would also have to parse the CSS to change the `url(...)`.

IMHO fontsource.org support is out of scope for this PR - it's not impossible but it sounds like it would be a lot of work and would increase the build size, which AFAICR was an underlying issue this PR was trying to reduce.

oetiker commented:
using fontsource with other js frameworks is exceedingly simple. 

```
import "@fontsource/open-sans/400.css";
import "@fontsource/open-sans/700.css";
```

this somehow causes that all the bits for these two font weights get copied during the build process.

but what goes on behind the scenes ... no idea.

hkollmann committed:
Merge branch 'master' into unified-webfonts

oetiker commented:
@johnspackman the tangible theme requires many of the material icons to work ... (https://github.com/johnspackman/qooxdoo/blob/unified-webfonts/source/class/qx/theme/tangible/Image.js) I guess that file needs some adapting ?

johnspackman committed:
fix explicit filenames in code

johnspackman committed:
Merge branch 'unified-webfonts' of github.com:johnspackman/qooxdoo into unified-webfonts

* 'unified-webfonts' of github.com:johnspackman/qooxdoo:
  fix preventDefault not working with autoDestroy (#10545)

johnspackman commented:
> @johnspackman the tangible theme requires many of the material icons to work ... (https://github.com/johnspackman/qooxdoo/blob/unified-webfonts/source/class/qx/theme/tangible/Image.js) I guess that file needs some adapting ?

oops, done

goldim commented on docs/desktop/gui/theming.md:
````diff
@@ -317,6 +305,60 @@ Note also the top level `weight`, `bold`, and `italic` properties - these do not
+
+The compiler can typically read the font and generate that mapping for you - eg:
+
+```
````
```suggestion
```bash
```

goldim commented on docs/desktop/gui/theming.md:
````diff
@@ -317,6 +305,60 @@ Note also the top level `weight`, `bold`, and `italic` properties - these do not
+By default, the compiler will use CDN URLs if they are given in Manifest.json, but you can override this in two ways: firstly, the `--local-fonts` compiler
+option, or adding a `fonts` object block in `compile.json`:
+
+```
````
```suggestion
```json5
```

goldim commented on test/framework/app/fonttest/source/class/fonttest/Application.js:
```diff
@@ -0,0 +1,73 @@
+     * 
+     * @lint ignoreDeprecated(alert)
+     */
+    main : function()
```
I would use ES5 syntax if it is template. Didn't get what it is.

goldim commented:
![image](https://user-images.githubusercontent.com/22189134/220940807-5691d2b7-0418-4806-b19e-ffaaa0467e73.png)

Compiled qxl.widgetbrowser and letter interval, line interval and etc were decreased in size.

goldim reviewed (changes requested):
I don't know why Widget Browser shows different things with this PR but think it should be fixed.

goldim commented on source/class/qx/tool/compiler/Analyser.js:
```diff
@@ -1415,6 +1419,47 @@ qx.Class.define("qx.tool.compiler.Analyser", {
+    /**
+     * Detects whether the filename is one of the fonts
+     *
+     * @param {*} filename
```
```suggestion
     * @param {String} filename
```

goldim commented on source/class/qx/tool/compiler/Analyser.js:
```diff
@@ -1415,6 +1419,47 @@ qx.Class.define("qx.tool.compiler.Analyser", {
+     * Detects whether the filename is one of the fonts
+     *
+     * @param {*} filename
+     * @returns
```
```suggestion
     * @returns {Boolean} whether the filename is a font asset
```

johnspackman committed:
Update docs/desktop/gui/theming.md

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

johnspackman committed:
Update docs/desktop/gui/theming.md

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

johnspackman committed:
Update docs/desktop/gui/theming.md

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

johnspackman committed:
Update source/class/qx/tool/compiler/Analyser.js

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

johnspackman committed:
Update source/class/qx/tool/compiler/Analyser.js

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

johnspackman committed:
bug fix

johnspackman committed:
Merge branch 'unified-webfonts' of github.com:johnspackman/qooxdoo into unified-webfonts

* 'unified-webfonts' of github.com:johnspackman/qooxdoo:
  Update source/class/qx/tool/compiler/Analyser.js
  Update source/class/qx/tool/compiler/Analyser.js
  Update docs/desktop/gui/theming.md
  Update docs/desktop/gui/theming.md
  Update docs/desktop/gui/theming.md

johnspackman committed:
removed unnecessary files

johnspackman commented:
> I don't know why Widget Browser shows different things with this PR but think it should be fixed.

done

johnspackman committed:
Merge branch 'master' into unified-webfonts

goldim commented:
Changing from Indigo -> any other theme -> Indigo in WidgetBrowser get next errors:
```
MProperty.js:67 Uncaught Error: No such property: EPSILON in qx.bom.Font (qx.bom.Font[821-0])
    at wrapper.set (MProperty.js:67:19)
    at wrapper._applyTheme (Font.js:236:26)
    at wrapper.set (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:3:1075)
    at wrapper.eval [as $$setThemeImpl] (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:6:258)
    at wrapper.eval [as setTheme] (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:3:21)
    at wrapper._applyTheme (Meta.js:81:15)
    at wrapper.set (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:3:1118)
    at wrapper.eval [as $$setThemeImpl] (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:4:258)
    at wrapper.eval [as setTheme] (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:3:21)
    at wrapper.<anonymous> (Header.js:74:45)
    at Direct.js:133:37
    at Function._true [as then] (Utils.js:149:22)
    at Direct.js:132:26
    at Array.forEach (<anonymous>)
    at wrapper.dispatchEvent (Direct.js:107:19)
    at wrapper.wrappedFunction [as dispatchEvent] (Interface.js:529:31)
```
```
Native.js:56 006077 qx.ui.core.queue.Layout: Error in the 'Layout' queue:TypeError: Cannot read properties of null (reading 'getInsets') TypeError: Cannot read properties of null (reading 'getInsets')
qx.ui.core.Widget:1314:33,qx.ui.core.Widget:1143:27,qx.ui.core.LayoutItem:566:48,qx.ui.core.queue.Layout:225:40,qx.ui.core.queue.Layout:89:26,qx.ui.core.queue.Manager:181:43,qx.ui.core.queue.Manager:233:11,qx.ui.core.queue.Manager:135:14,qx.event.dispatch.Direct:123:39,qx.event.Utils:190:24,qx.event.dispatch.Direct:122:28,Array.forEach (<anonymous>),qx.event.dispatch.Direct:110:21,qx.Interface:514:33,qx.event.Manager:911:52,qx.event.Registration:350:42
```
```
Native.js:56 006749 qx.ui.core.queue.Appearance: Error in the 'Appearance' queue:Error: Error in property decorator of class qx.ui.form.ListItem in method setThemedDecorator with incoming value 'selected': Is invalid! Error: Error in property decorator of class qx.ui.form.ListItem in method setThemedDecorator with incoming value 'selected': Is invalid!
qx.core.Property:773:15,qx.core.Property:820:38,qx.core.Property:820:38,qx.ui.core.Widget:2443:86,qx.ui.core.queue.Appearance:104:17,qx.ui.core.queue.Manager:165:47,qx.ui.core.queue.Manager:233:11,qx.ui.core.queue.Manager:135:14,qx.event.dispatch.Direct:123:39,qx.event.Utils:190:24,qx.event.dispatch.Direct:122:28,Array.forEach (<anonymous>),qx.event.dispatch.Direct:110:21,qx.Interface:514:33,qx.event.Manager:911:52,qx.event.Registration:350:42
```

goldim commented:
```
Uncaught Error: Error in property decorator of class qx.ui.form.ListItem in method setThemedDecorator with incoming value 'selected': Is invalid!
    at Object.error (Property.js:943:13)
    at wrapper.set (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:3:773)
    at wrapper.eval [as setThemedDecorator] (eval at __installFunctionFromCode__P_57_13 (Property.js:1001:36), <anonymous>:6:258)
    at wrapper.syncAppearance (Widget.js:2463:15)
    at wrapper.removeState (Widget.js:2300:14)
    at wrapper._onPointerOut (ListItem.js:99:12)
    at EventHandler.js:298:35
    at Function._true [as series] (Utils.js:318:24)
    at EventHandler.js:296:31
    at Function._true [as then] (Utils.js:149:22)
    at wrapper._dispatchEvent (EventHandler.js:295:22)
    at AbstractBubbling.js:260:39
    at Function._true [as series] (Utils.js:318:24)
    at AbstractBubbling.js:228:39
    at Function._true [as series] (Utils.js:318:24)
    at AbstractBubbling.js:225:31
```
```
Native.js:56 006788 qx.ui.core.queue.Appearance: Error in the 'Appearance' queue:Error: Error in property decorator of class qx.ui.form.ListItem in method setThemedDecorator with incoming value 'selected': Is invalid! Error: Error in property decorator of class qx.ui.form.ListItem in method setThemedDecorator with incoming value 'selected': Is invalid!
qx.core.Property:773:15,qx.core.Property:820:38,qx.core.Property:820:38,qx.ui.core.Widget:2443:86,qx.ui.core.queue.Appearance:104:17,qx.ui.core.queue.Manager:165:47,qx.ui.core.queue.Manager:233:11,qx.ui.core.queue.Manager:135:14,qx.event.dispatch.Direct:123:39,qx.event.Utils:190:24,qx.event.dispatch.Direct:122:28,Array.forEach (<anonymous>),qx.event.dispatch.Direct:110:21,qx.Interface:514:33,qx.event.Manager:911:52,qx.event.Registration:350:42
```

goldim commented:
![image](https://user-images.githubusercontent.com/22189134/220967898-24ff44c4-2e9f-4843-af43-9861b18f5f2b.png)

Get `...` even in fullscreen mode
Changing to TangiableLight:
![image](https://user-images.githubusercontent.com/22189134/220968389-a3f6104b-5627-484b-87cd-b539350fee60.png)

And text like cut at the bottom

goldim commented on source/class/qx/tool/compiler/Analyser.js:
```diff
@@ -1415,6 +1419,47 @@ qx.Class.define("qx.tool.compiler.Analyser", {
+     *
+     * @param {String} name
+     * @param {Boolean?} create whether to create the font if it does not exist (default is false)
+     * @returns {qx.tool.app.ManifestFont?} null if it does not exist and `create` is falsey
```
```suggestion
     * @returns {qx.tool.app.ManifestFont?} null if it does not exist and `create` is false
```

johnspackman replied:
it's falsey not false, so the comment is correct

goldim commented on source/class/qx/tool/compiler/app/Application.js:
```diff
@@ -833,6 +833,25 @@ qx.Class.define("qx.tool.compiler.app.Application", {
     },
 
+    /**
+     * Returns a list of fonts required by the application, where the font is detailed in Manfest.json
```
```suggestion
     * Returns a list of fonts required by the application, where the font is detailed in Manifest.json
```

goldim commented on source/class/qx/tool/compiler/app/Application.js:
```diff
@@ -833,6 +833,25 @@ qx.Class.define("qx.tool.compiler.app.Application", {
+     * @returns {String[]}
+     */
+    getFonts() {
+      var fonts = {};
```
```suggestion
      const fonts = {};
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/app/Application.js:
```diff
@@ -833,6 +833,25 @@ qx.Class.define("qx.tool.compiler.app.Application", {
+     */
+    getFonts() {
+      var fonts = {};
+      var analyser = this.getAnalyser();
```
```suggestion
      const analyser = this.getAnalyser();
```

goldim commented on source/class/qx/tool/compiler/app/Application.js:
```diff
@@ -833,6 +833,25 @@ qx.Class.define("qx.tool.compiler.app.Application", {
+      var analyser = this.getAnalyser();
+      var db = analyser.getDatabase();
+      this.__loadDeps.forEach(classname => {
+        var classInfo = db.classInfo[classname];
```
```suggestion
        const classInfo = db.classInfo[classname];
```

goldim commented on source/class/qx/tool/compiler/app/Library.js:
```diff
@@ -366,6 +385,23 @@ qx.Class.define("qx.tool.compiler.app.Library", {
+     * Detects whether the filename is one of the library's fonts
+     *
+     * @param {String} filename
+     * @returns
```
```suggestion
     * @returns {Boolean}
```

goldim commented on source/class/qx/tool/compiler/app/ManifestFont.js:
```diff
@@ -0,0 +1,171 @@
+     * Updates this from the data in the Manifest.json
+     *
+     * @param {Object} data the data from Manifest
+     * @param {qx.tool.compiler.app.Library} library the library
```
```suggestion
     * @param {qx.tool.compiler.app.Library} the library
```

goldim commented on source/class/qx/tool/utils/Http.js:
```diff
@@ -0,0 +1,69 @@
+/**
+ * Helper methods for HTTP
+ */
+qx.Class.define("qx.tool.utils.Http", {
```
Maybe should this class have a `static` type?

goldim commented on source/class/qx/theme/manager/Font.js:
```diff
@@ -216,6 +216,14 @@ qx.Class.define("qx.theme.manager.Font", {
             this.__resolveInclude(source, key);
           }
 
+          if (source[key].fontName) {
+            let preset = this._manifestFonts[source[key].fontName];
+            Object.keys(preset).forEach(key => {
+              if (source[key] === undefined) {
+                source[key] = preset[key];
+              }
+            });
+          }
```
```suggestion
          const fontName = source[key].fontName;
          if (fontName) {
            let preset = this._manifestFonts[fontName];
            Object.keys(preset)
              .filter(key => source[key] === undefined)
              .forEach(key => source[key] = preset[key]);
          }
```

goldim commented on source/class/qx/tool/compiler/ClassFile.js:
```diff
@@ -631,6 +636,22 @@ qx.Class.define("qx.tool.compiler.ClassFile", {
       }
+
+      // Fonts
+      var fontNames = Object.keys(this.__requiredFonts);
```
```suggestion
      let fontNames = Object.keys(this.__requiredFonts);
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/ClassFile.js:
```diff
@@ -2896,6 +2922,18 @@ qx.Class.define("qx.tool.compiler.ClassFile", {
 
+    /**
+     * Adds a required font
+     * @param {String} name  name of the font
```
Is duplicating name fine? Noticed it in several places of PR.

derrell replied:
The second `name` could be capitalized to make it more clear.

I tend to format my JSDoc with the description on a new line, so that problem doesn't show up. I would have done this:
```
@param {String} name
    name of the font
```

goldim replied:
Got it, thank you

goldim commented on source/class/qx/tool/compiler/app/ManifestFont.js:
```diff
@@ -0,0 +1,171 @@
+    this.setName(name);
+  },
+
+  properties: {
```
No JSDoc description for properties

johnspackman replied:
what are you proposing?  That it says:

```
/* these are properties */
properties: {
```

Do we have that kind of comment in every other file??

goldim replied:
I try to explain. The class `qx.tool.compiler.app.ManifestFont` has `properties` section and every property without JSDoc comment section above it (Properties ). Ofc, I could write the same phrase for each property but does it worth it? I remember at one of my PR I was asked to add jsdoc for properties.

goldim commented on source/class/qx/tool/compiler/app/ManifestFont.js:
```diff
@@ -0,0 +1,171 @@
+        "sources",
+        "glyphs",
+        "family"
+      ].forEach(name => {
+        if (data.hasOwnProperty(name)) {
+          toSet[name] = data[name];
+        }
+      });
```
```suggestion
      ].filter(name => data.hasOwnProperty(name))
      .forEach(name => toSet[name] = data[name]);
```

johnspackman replied:
this is introducing a bug, and the reason why I wrote it as it was and not with `Object.assign`

goldim commented on source/class/qx/tool/compiler/targets/Target.js:
```diff
@@ -506,88 +522,161 @@ qx.Class.define("qx.tool.compiler.targets.Target", {
       var assetUris = application.getAssetUris(t, rm, appMeta.getEnvironment()); // Save any changes that getAssets collected
```
```suggestion
      let assetUris = application.getAssetUris(t, rm, appMeta.getEnvironment()); // Save any changes that getAssets collected
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/targets/Target.js:
```diff
@@ -506,88 +522,161 @@ qx.Class.define("qx.tool.compiler.targets.Target", {
+      await this._writeTranslations();
 
-      var fontCntr = 0;
       var assets = {};
```
```suggestion
      let assets = {};
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/targets/Target.js:
```diff
@@ -506,88 +522,161 @@ qx.Class.define("qx.tool.compiler.targets.Target", {
           }
-        };
+        }
+        var code = font.getBootstrapCode(
```
```suggestion
        const code = font.getBootstrapCode(
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/targets/Target.js:
```diff
@@ -506,88 +522,161 @@ qx.Class.define("qx.tool.compiler.targets.Target", {
+          return;
+        }
+        let resources = font.getApplicationFontData();
+        for (var key in resources) {
```
```suggestion
        for (let key in resources) {
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/targets/Target.js:
```diff
@@ -506,88 +522,161 @@ qx.Class.define("qx.tool.compiler.targets.Target", {
+      let appLibrary = appMeta.getAppLibrary();
+      let fontsToLoad = {};
+      const addLibraryFonts = library => {
+        var fonts = library.getWebFonts();
```
```suggestion
        let fonts = library.getWebFonts();
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/targets/Target.js:
```diff
@@ -506,88 +522,161 @@ qx.Class.define("qx.tool.compiler.targets.Target", {
-                application.getName(),
-                font.getResources().join(",")
-              );
+          var code = font.getBootstrapCode(this, application);
```
```suggestion
          const code = font.getBootstrapCode(this, application);
```

johnspackman replied:
not necessary

goldim commented on source/class/qx/tool/compiler/targets/Target.js:
```diff
@@ -506,88 +522,161 @@ qx.Class.define("qx.tool.compiler.targets.Target", {
+          return;
+        }
+        let resources = font.getApplicationFontData();
+        for (var key in resources) {
```
```suggestion
        for (let key in resources) {
```

johnspackman replied:
not necessary

johnspackman committed:
Update source/class/qx/tool/compiler/app/Application.js

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

johnspackman commented on source/class/qx/tool/compiler/app/Application.js:
```diff
@@ -833,6 +833,25 @@ qx.Class.define("qx.tool.compiler.app.Application", {
+    getFonts() {
+      var fonts = {};
+      var analyser = this.getAnalyser();
+      var db = analyser.getDatabase();
```
not necessary

johnspackman committed:
Update source/class/qx/tool/compiler/app/Library.js

Co-authored-by: Dmitry Zolotov <goldim1993@mail.ru>

johnspackman commented:
> Changing from Indigo -> any other theme -> Indigo in WidgetBrowser get next errors:

fixed

> Uncaught Error: Error in property decorator of class qx.ui.form.ListItem in method setThemedDecorator with incoming value 'selected': Is invalid!

I cant reproduce this but it looks like it would be a bug in a theme, ie missing decorator

> Get ... even in fullscreen mode

fixed

johnspackman committed:
bugfixes

johnspackman committed:
Merge branch 'unified-webfonts' of github.com:johnspackman/qooxdoo into unified-webfonts

* 'unified-webfonts' of github.com:johnspackman/qooxdoo:
  Update source/class/qx/tool/compiler/app/Library.js
  Update source/class/qx/tool/compiler/app/Application.js

hkollmann committed:
Merge branch 'master' into unified-webfonts

goldim commented:
Checked on 3 apps everything works fine,

hkollmann commented:
@derrell could you approve?

derrell commented:
I've already approved this. It's @level420 who this one's waiting on.

hkollmann commented:
What's with your change request?

goldim reviewed (changes requested):
On qooxdoo site
![image](https://user-images.githubusercontent.com/22189134/221375570-a1d40183-77ee-4cfc-89bf-8073a5216078.png)

And now at bottom button text:

![image](https://user-images.githubusercontent.com/22189134/221375614-e6eb623e-c6fa-4bff-94b3-635fcc67f577.png)

goldim commented:
Also text of Widget Browser on site qooxdoo (online version) is wider than in Widget Browser which uses PR compiler.

hkollmann commented:
> Also text of Widget Browser on site qooxdoo (online version) is wider than in Widget Browser which uses PR compiler.

Could this be a ignorable rounding issue? The difference is marginal and all pixel based text are working.

hkollmann committed:
Merge branch 'master' into unified-webfonts

hkollmann commented:
@goldim : Could you agree?

goldim commented:
@hkollmann tbh I dunno. I would like to know what @johnspackman thinks about it. I am not against this PR would be merged if It will be fixed before new release. But if doesnt I predict some user complains about it.

goldim commented:
Especially as It was with hurried removal of MaterialIcons.

oetiker commented:
> > Also text of Widget Browser on site qooxdoo (online version) is wider than in Widget Browser which uses PR compiler.
> 
> Could this be a ignorable rounding issue? The difference is marginal and all pixel based text are working.

as the bottom lower case g gets cut of, I don't think this is acceptable and I am pretty sure this indicates a bug somewhere ... in any case if I was going to ship an application where the letters of the font get cut off, this would not pass muster ...

johnspackman commented:
it's a bug, ill take a look now

johnspackman committed:
fix font display

johnspackman committed:
Merge branch 'unified-webfonts' of github.com:johnspackman/qooxdoo into unified-webfonts

* 'unified-webfonts' of github.com:johnspackman/qooxdoo:
  Removed nonexistent pages: remote_io and rpc. Added Transport Layer page. Fixed broken pages in menu. (#10552)
  Since 5.0 version of sinon, sinon object is default sandbox and sandbox and createSandbox were removed. (#10547)

goldim commented:
@johnspackman the issue was gone but still in indigo theme text looks like compressed horizontally. Something like forgotten for Indigo theme. Can you see it? I could send screen record showing this problem.

johnspackman commented:
> The bottom portion of the lower case g gets cut off, I don't think this is acceptable.

fixed

> Especially as It was with hurried removal of MaterialIcons.

nothing's been removed, this PR is about making the loading order/mechanism much finer grained; the font in use is actually Roboto

> the issue was gone but still in indigo theme text looks like compressed horizontally. Something like forgotten for Indigo theme. Can you see it? I could send screen record showing this problem.

I've checked the button fonts are they are definitely the same size in indigo on mine - yes please for a screenshot

> And also in tangiable themes the text became bigger by 2px.

I definitely can't see this, when I get it the font sizes are the same - but I have overlaid a couple of screenshots on top of each other and can see that there are some differences in layout, and I seen another group title with ellipsis too :(

johnspackman committed:
first pass at fixing font loading cdn & local

johnspackman committed:
bugfix

johnspackman commented:
hopefully this is the last fix; there is still a difference between the widget browser produced by this version compared to the old version, BUT it only happens if we use the Roboto fonts locally and not via a CDN (ie this affects Tangible theme).

AIUI there may be a change in the font definition, but also the CDN delivers multiple fonts (eg same fontFamily but different fontWeight and fontStyle) in a single CSS - this is the reason why there is such a change in the last push, it's because the previous loader was assuming that only a single font in a single family could be loaded.

johnspackman commented:
PS hang on before merging please I just remembered I need to update the documentation! :(

johnspackman committed:
no message

johnspackman commented:
docs are done now

johnspackman committed:
delete rogue console.log

johnspackman commented:
@goldim oops, must have copied & pasted!  fixed the fonts for indigo now

johnspackman committed:
wrong fonts for indigo

hkollmann commented:
@johnspackman: test are not working any longer?

johnspackman committed:
fix unit tests

johnspackman committed:
another fix

johnspackman committed:
another fix

johnspackman commented:
there is one test which is still failing - but I can't reproduce that on my Mac with Chrome, Firefox, or Safari, or on Fedora 36 with Chromium or Firefox.

Before I install yet another linux VM, does anyone have a Ubuntu VM handy they can test this with?

goldim commented:
@johnspackman I could run tests on Debian locally.

johnspackman commented:
Yes please; not sure what else to try except reproduce the GitHub environment exactly :(

goldim commented:
It's pity but sometimes Debian behavior differs from Ubuntu

johnspackman committed:
eliminate the test - it is faulty and not testing anything really

hkollmann committed:
Merge branch 'master' into unified-webfonts

hkollmann committed:
Merge branch 'master' into unified-webfonts
